worker/src/lib.rs

```diff
@@ -0,0 +1,255 @@
+                if let Err(e) = ops::stop_child(&mut child).await {
+                    error!("{:?}", e);
+                }
+            },
+            res = child.wait() => {
+                match res {
+                    Ok(s) => {
+                        inner.finish(s).await;
+                        inner.progress.notify_waiters();
+                        return;
+                    }
+                    Err(_) => unimplemented!("when does this happen?"), // TODO
+                }
+            }
+        }
+    }
+}
+
+pub(crate) struct ProcessInner {
+    exit_status: RwLock<Option<ExitStatus>>,
+    logs: RwLock<Vec<Bytes>>,
+
+    // Signals the listeners about progress being made by the process
+    // (either new log messages or finishing).
+    progress: Arc<Notify>,
+
+    stop_sender: Mutex<Option<oneshot::Sender<()>>>,
+}
+
+impl ProcessInner {
+    fn new(stop_sender: oneshot::Sender<()>) -> Self {
+        Self {
+            exit_status: Default::default(),
+            logs: Default::default(),
+            progress: Default::default(),
+            stop_sender: Mutex::new(Some(stop_sender)),
+        }
+    }
+    async fn finish(&self, exit_status: ExitStatus) {
+        *self.exit_status.write().await = Some(exit_status);
+    }
+}
+
+/// Represents a single process.
+pub struct Process(Arc<ProcessInner>);
+
+impl Process {
+    /// Spawn a new process.
+    pub fn spawn<'a>(argv0: &str, argv: impl Iterator<Item = &'a str>) -> Result<Self, io::Error> {
+        let (stop_tx, stop_rx) = oneshot::channel();
+        let inner = Arc::new(ProcessInner::new(stop_tx));
+        let inner_clone = inner.clone();
+        let child = Command::new(argv0)
+            .args(argv)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()?;
+
+        tokio::spawn(process_task(child, inner, stop_rx));
+
+        Ok(Process(inner_clone))
+    }
+
+    /// Tries to stop the process, first gracefully by sending SIGTERM,
+    /// then, after timeout specified by `SIGTERM_TIMEOUT` elapses, by sending a SIGKILL.
+    /// If the process has already finished, returns `Ok` with the exit status.
+    /// If the process has not finished,
+    /// but another "stop" operation has already been initiated, returns `Err(())`.
+    pub async fn stop(&self) -> Result<ExitStatus, ()> {
+        if let Some(e) = *self.0.exit_status.read().await {
+            return Ok(e);
+        }
+
+        match self.0.stop_sender.lock().await.take() {
+            Some(tx) => {
+                let notify = self.0.progress.clone();
+
+                // Ignore error: if receiver has hung up, process has already finished.
+                let _ = tx.send(());
+
+                loop {
+                    let notified = notify.notified();
+                    if let Some(e) = *self.0.exit_status.read().await {
+                        return Ok(e);
+                    }
+                    notified.await;
```

Comment:
I'm probably just missing something since I don't have very deep exposure to async/await, but why not just do this as a oneliner `notify.notified().await`

Reply:
So, [notify_waiters](https://docs.rs/tokio/1.3.0/tokio/sync/struct.Notify.html#method.notify_waiters) (which signals this) only notifies the waiters that have already registered using `notified()`. I am pretty sure that if we did:
```rust
loop {
    if exited {
        return;
    }
    notify.notified().await
}
```

, this race condition is possible:
1. T1 tries to read exit status, it is not available yet
2. T2 writes exit status, notifies waiters
3. T1 executes `notified().await`, i.e. registers interest in notifications and waits for one. The notification has already been missed.

T = task or thread (Tokio is multi-threaded by default AFAIK)

Reply:
Your reasoning @justinas is correct based on my knowledge and experience.

Reply:
This is a fairly common error I see when `Notify` is used.

Reply:
got it, awesome, thanks.